server/remote_cache/scorecard/scorecard_test.go

```diff
@@ -0,0 +1,210 @@
+			Mask:   &fieldmaskpb.FieldMask{Paths: []string{"search"}},
+			Search: "bes-upload",
+		},
+	}
+
+	res, err := scorecard.GetCacheScoreCard(ctx, env, req)
+	require.NoError(t, err)
+
+	assertResults(t, res, besUpload)
+}
+
+func TestGetCacheScoreCard_Filter_CacheType(t *testing.T) {
+	ctx := context.Background()
+	env := setupEnv(t)
+	req := &capb.GetCacheScoreCardRequest{
+		Filter: &capb.GetCacheScoreCardRequest_Filter{
+			Mask:      &fieldmaskpb.FieldMask{Paths: []string{"cache_type"}},
+			CacheType: capb.CacheType_AC,
+		},
+	}
+
+	res, err := scorecard.GetCacheScoreCard(ctx, env, req)
+	require.NoError(t, err)
+
+	assertResults(t, res, acMiss)
+}
+
+func TestGetCacheScoreCard_Filter_RequestType(t *testing.T) {
+	ctx := context.Background()
+	env := setupEnv(t)
+	req := &capb.GetCacheScoreCardRequest{
+		Filter: &capb.GetCacheScoreCardRequest_Filter{
+			Mask:        &fieldmaskpb.FieldMask{Paths: []string{"request_type"}},
+			RequestType: capb.RequestType_READ,
+		},
+	}
+
+	res, err := scorecard.GetCacheScoreCard(ctx, env, req)
+	require.NoError(t, err)
+
+	assertResults(t, res, acMiss, casDownload)
+}
+
+func TestGetCacheScoreCard_Filter_ResponseType(t *testing.T) {
+	ctx := context.Background()
+	env := setupEnv(t)
+	req := &capb.GetCacheScoreCardRequest{
+		Filter: &capb.GetCacheScoreCardRequest_Filter{
+			Mask:         &fieldmaskpb.FieldMask{Paths: []string{"response_type"}},
+			ResponseType: capb.ResponseType_NOT_FOUND,
+		},
+	}
+
+	res, err := scorecard.GetCacheScoreCard(ctx, env, req)
+	require.NoError(t, err)
+
+	assertResults(t, res, acMiss)
+}
+
+func TestGetCacheScoreCard_Sort_StartTime(t *testing.T) {
+	ctx := context.Background()
+	env := setupEnv(t)
+	req := &capb.GetCacheScoreCardRequest{
+		OrderBy: capb.GetCacheScoreCardRequest_ORDER_BY_START_TIME,
+	}
+
+	res, err := scorecard.GetCacheScoreCard(ctx, env, req)
+	require.NoError(t, err)
+
+	assertResults(t, res, besUpload, casUpload, acMiss, casDownload)
+}
+
+func TestGetCacheScoreCard_Sort_Duration(t *testing.T) {
+	ctx := context.Background()
+	env := setupEnv(t)
+	req := &capb.GetCacheScoreCardRequest{
+		OrderBy: capb.GetCacheScoreCardRequest_ORDER_BY_DURATION,
+	}
+
+	res, err := scorecard.GetCacheScoreCard(ctx, env, req)
+	require.NoError(t, err)
+
+	assertResults(t, res, acMiss, casDownload, casUpload, besUpload)
+}
+
+func TestGetCacheScoreCard_GroupByActionOrderByDurationDesc(t *testing.T) {
+	ctx := context.Background()
+	env := setupEnv(t)
+	req := &capb.GetCacheScoreCardRequest{
+		OrderBy:       capb.GetCacheScoreCardRequest_ORDER_BY_DURATION,
+		GroupByAction: true,
+		Descending:    true,
+	}
+
+	res, err := scorecard.GetCacheScoreCard(ctx, env, req)
+	require.NoError(t, err)
+
+	assertResults(t, res, besUpload, casUpload, acMiss, casDownload)
+}
+
+func assertResults(t *testing.T, res *capb.GetCacheScoreCardResponse, msg ...*capb.ScoreCard_Result) {
+	// Note: not asserting directly on the protos because the diff is too hard to read.
```

Comment:
take a look at protocmp and consider whether it's useful here: https://github.com/buildbuddy-io/buildbuddy/blob/4d9c9965dcfe967822a1269648f95c976c263fda/enterprise/server/workflow/service/service_test.go#L289

Reply:
Thanks for pointing me to that -- I didn't know about go-cmp before, and it looks super useful.

I took a look at the diff that it outputs, and it has the same problem as the testify diff -- it does field-level diffs which are a bit too fine-grained for this particular test where the protos just being out of order causes a lot of diffs. Ideally, it would say something like "these lists have the same protos, but in the wrong order," but alas.